components/tidb_query/src/codec/mysql/time/datetime.rs

```diff
@@ -0,0 +1,1243 @@
+
+        if config.strict_mode {
+            sql_mode |= SqlMode::STRICT_ALL_TABLES;
+        }
+        if config.allow_invalid_date {
+            sql_mode |= SqlMode::INVALID_DATES;
+        }
+        if config.no_zero_date {
+            sql_mode |= SqlMode::NO_ZERO_DATE;
+        }
+        if config.no_zero_in_date {
+            sql_mode |= SqlMode::NO_ZERO_IN_DATE;
+        }
+
+        if config.ignore_truncate {
+            flags |= Flag::IGNORE_TRUNCATE;
+        }
+
+        eval_config.set_sql_mode(sql_mode).set_flag(flags);
+
+        EvalContext::new(Arc::new(eval_config))
+    }
+}
+
+#[derive(PartialEq, Clone, Copy, Debug)]
+pub enum TimeType {
+    Date,
+    DateTime,
+    TimeStamp,
+}
+
+impl TryFrom<FieldTypeTp> for TimeType {
+    type Error = codec::Error;
+    fn try_from(time_type: FieldTypeTp) -> Result<TimeType> {
+        Ok(match time_type {
+            FieldTypeTp::Date => TimeType::Date,
+            FieldTypeTp::DateTime => TimeType::DateTime,
+            FieldTypeTp::Timestamp => TimeType::TimeStamp,
+            _ => return Err(box_err!("Time does not support field type {}", time_type)),
+        })
+    }
+}
+
+// The common set of methods for `date/time`
+impl Time {
+    /// Returns the hour number from 0 to 23.
+    #[inline]
+    pub fn hour(self) -> u32 {
+        self.get_hour()
+    }
+
+    /// Returns the minute number from 0 to 59.
+    #[inline]
+    pub fn minute(self) -> u32 {
+        self.get_minute()
+    }
+
+    /// Returns the second number from 0 to 59.
+    #[inline]
+    pub fn second(self) -> u32 {
+        self.get_second()
+    }
+
+    /// Returns the number of microseconds since the whole second.
+    pub fn micro(self) -> u32 {
+        self.get_micro()
+    }
+
+    /// Returns the year number
+    pub fn year(self) -> u32 {
+        self.get_year()
+    }
+
+    /// Returns the month number
+    pub fn month(self) -> u32 {
+        self.get_month()
+    }
+
+    /// Returns the day number
+    pub fn day(self) -> u32 {
+        self.get_day()
+    }
+}
+
+// Parser
+impl Time {
+    // Split `input` with `.`, returns the whole part and a optional fractional part.
+    fn split_frac(input: &[u8]) -> Result<(&[u8], Option<&[u8]>)> {
+        let parts = input.split(|&x| x == b'.').collect::<Vec<_>>();
+
+        if parts.len() > 2 {
+            Err(Error::truncated())
+        } else {
+            Ok((parts[0], parts.get(1).copied()))
+        }
+    }
+
+    // Split the datetime with whitespace or 'T'.
+    fn split_datetime(input: &[u8]) -> Vec<&[u8]> {
+        input
+            .split(|&x| x.is_ascii_whitespace() || x == b'T')
+            .collect::<Vec<_>>()
+    }
+
+    fn is_float_date(input: &[u8]) -> bool {
+        input.iter().all(|&x| x == b'.' || x.is_ascii_digit())
+    }
+
+    fn round(parts: &mut [u32]) -> Result<()> {
+        if parts.len() != 7 {
+            return Err(Error::truncated());
+        }
+        let modulus = [
```

Comment:
How about make `modules` array global. Then we needn't to allocate this space and init it every time we call this function

Reply:
In fact, we can't. We never know the last day of the month until we get the `year` and `month`

Reply:
we can define a global private array
```rust
        const modulus = [
            0,
            12,
            std::u32::MAX
            24,
            60,
            60,
            1_000_000,
        ];
```
then, split the cycle using this array to two part, one is before `std::u32::MAX`, one is  after `std::u32::MAX`. 
And, add enough doc for how to use this array.
Why there is `std::u32::MAX`, if the user use this array mistakenly, then it will panic(just like, `10 as u32 - u32::MAX` will panic)

Reply:
@iosmanthus ping.
If the `moduls` is local var, it will be construct every time we call this function.
Make it be global const array and use split cycle in this function to use it.